### PR TITLE
fix: remove trivial allocations and `try_into`

### PR DESCRIPTION
**Pull request by Daniel-Aaron-Bloom**
Use the following array tools to avoid allocations and `try_into` calls:

- `core::array::from_fn`/`core::array::try_from_fn`
- `[T]::first_chunk`
- `[T]::array_chunks`
- `[T; N]::map`
- `iter_fixed::IntoIteratorFixed`

Some of the `try_into` calls are more of a nit just because they look like allocations but are really just slice conversions, but it was easier to just get rid of all of them.

### CODE REVIEW

**Review by Nashtare — COMMENTED** [review]
As an overall comment, we've been working on removing unstable features from these libraries (with the current status of having `specialization` remaining for `plonky2_field`).

This PR isn't really aligned in that it brings two additional unstable features, and as such, I am not enclined in merging this. As a general rule, such addition should be justified by significant performance improvement to balance out the drawback.

**Comment by Daniel-Aaron-Bloom** [comment]
Fair.

I would note though that `specialization` is probably the least likely feature to be stabilized in the near future, so depending on your plans to remove that from `plonky2_field`, the introduction of these features may or may not lengthen the timeline to stability (assuming that's your goal).

Other than micro-benchmarks, but there's no real perf impact of this change. I can trim it down to just the stable if there's any interest in that.

**Comment by muursh** [comment]
> I would note though that `specialization` is probably the least likely feature to be stabilized in the near future, so depending on your plans to remove that from `plonky2_field`, the introduction of these features may or may not lengthen the timeline to stability (assuming that's your goal).

I'd rather avoid adding more unstable features no matter what tbh.

> Other than micro-benchmarks, but there's no real perf impact of this change. I can trim it down to just the stable if there's any interest in that.

I'm open to that.

File: plonky2/src/hash/hash_types.rs [hunk 1]
```diff
@@ -90,14 +90,11 @@ impl<F: RichField> GenericHashOut<F> for HashOut<F> {
     }
 
     fn from_bytes(bytes: &[u8]) -> Self {
+        // TODO: replace with as_chunks
```
Comment (Nashtare) [inline plonky2/src/hash/hash_types.rs:93]: Same remark as the other TODO

File: plonky2/src/iop/ext_target.rs [hunk 2]
```diff
@@ -150,9 +151,15 @@ pub fn flatten_target<const D: usize>(l: &[ExtensionTarget<D>]) -> Vec<Target> {
 }
 
 /// Batch every D-sized chunks into extension targets.
-pub fn unflatten_target<const D: usize>(l: &[Target]) -> Vec<ExtensionTarget<D>> {
+pub fn unflatten_target<const D: usize>(mut l: &[Target]) -> Vec<ExtensionTarget<D>> {
     debug_assert_eq!(l.len() % D, 0);
-    l.chunks_exact(D)
-        .map(|c| c.to_vec().try_into().unwrap())
-        .collect()
+    // TODO: replace with array_chunks
```
Comment (Nashtare) [inline plonky2/src/iop/ext_target.rs:156]: Same remark

**Review by Nashtare — COMMENTED** [review]
I also have concerns about the introduction of another dependency, `iter_fixed`, which hasn't been maintained for the past 3 years.

**Comment by Nashtare** [comment]
Closing for inactivity.

**Pull request closed without merging**